README.md
# Imperfect

Because everyone's software runs in the real world.

It's an unfortunate fact of life that not everything always works
perfectly. However more unfortunately is that all the error reporting
solutions assume that there is a constant threshold above which an error
becomes an alert.

Imperfect takes a different approach. Imperfect measures the ratio of
failures over a given period of time, which is much more useful as you
really care about your success rate rather than an the absolute number
of failures.

## Integrations.

* For storage it uses Amazon Cloudwatch.
* For alerting it uses Pagerduty.

## Setup

1. `gem install imperfect`
1. Configure it.
1. Call it using `Imperfect.success('event')` and `Imperfect.failure('event')`
   to approriately track the event you'd like to watch.

If you'd like to trigger a failure for testing purposes you can call
`Imperfect.trigger_alert!('event')` and it will send an alert using the
current configuration (NOTE: this is to be implemented).

## Configuration

```
Imperfect.configure do |config|
  config.enabled = Rails.env.production?

  config.enable_storage :cloudwatch, {
    :access_key_id => "your-aws-id",
    :secret_access_key => "your-aws-secret"
  }
  config.enable_alerts :pagerduty, {
    :api_key => "your-pagerduty-api-key"
  }

  # Minimum time (in seconds) between re-triggering alerts
  config.minimum_alert_update_interval => 60

  # All events that you would alert on must be specified here. Unconfigured
  # events will be silently ignored.
  config.events = {
    'event' => {
      # Alert if the failure rate exceeds this value.
      :acceptable_failure_rate => 0.10,

      # The period of preceding time used to determine the current failure rate,
      # defaults to 300s.
      :lookback_period => 300,

      :alerts => {
        :pagerduty => {
          # The name of the pagerduty service to mark as failed.
          :service_name => 'service-event',
        }
      },

      :storage => {
        # The group which this metric should belong to on cloudwatch.
        :cloudwatch_namespace => 'Imperfect',
        # The name of the cloudwatch metric to use.
        :cloudwatch_metric_name => 'event',
      }
    }
  }
end
```

# Development

1. Checkout the repository
1. Create a feature branch `git checkout -b my-new-feature`
1. `bundle install`
1. Run `guard`
1. Submit a pull request (pulls should be mergable into master, include
   tests for new functionality and be currently passing).

# Future Work

1. Alerting (stub is there but needs implementation)
1. Other data providers (graphite?)
1. Other alert services


CHANGELOG.md

0.0.3 / 2014-02-24
==================

 * Update dependencies
 * Update license

0.0.1 / YYYY-MM-DD
==================

  * Initial version


spec/spec_helper.rb
require 'bundler/setup'
require 'imperfect'

require 'webmock/rspec'

RSpec.configure do |config|
end


spec/lib/imperfect_spec.rb
require 'spec_helper'

describe 'Imperfect' do
  subject do
    Imperfect
  end

  it 'should have a version' do
    Imperfect::VERSION.should_not be_nil
  end

  describe 'configuring' do
    context 'global enabled' do
      before do
        subject.configure do |config|
          config.enabled = true
        end
      end
      it 'exists' do
        subject.configuration.enabled.should be_true
      end
    end

    context 'storage with cloudwatch' do
      before do
        subject.configure do |config|
          config.enable_storage(:cloudwatch, :access_key_id => 'id', :secret_key => 'secret')
        end
      end
      it 'exists' do
        subject.configuration.storage.should_not be_nil
      end
    end

    context 'alerting with pagerduty' do
      before do
        subject.configure do |config|
          config.enable_alerts(:pagerduty, :api_key => 'secret')
        end
      end
      it 'exists' do
        subject.configuration.alerts.should_not be_nil
      end
    end

    context 'configuring an event' do
      before do
        subject.configure do |config|
          config.events = {
            'event' => {
              :acceptable_failure_rate => 0.10,
              :lookback_period => 300,
              :alerts => {
                :pagerduty => {
                  :service_name => 'service-event',
                }
              },
              :storage => {
                :cloudwatch_namespace => 'Imperfect',
                :cloudwatch_metric_name => 'event',
              }
            }
          }
        end
      end

      it 'has an event' do
        subject.configuration.events['event'].should_not be_nil
      end
    end
  end

  context 'with storage' do
    let(:event_configuration) {
      {
        :storage => {
          :cloudwatch_namespace => 'Imperfect',
          :cloudwatch_metric_name => 'event',
        }
      }
    }

    before do
      subject.configure do |config|
        config.enable_storage(:cloudwatch, :access_key_id => 'id', :secret_key => 'secret')
        config.events = { 'event' => event_configuration }
      end
    end

    context 'when enabled' do
      describe 'success' do
        it 'stores a datapoint' do
          subject.configuration.storage.should_receive(:update).with(event_configuration[:storage], 'event', :success)
          subject.success('event')
        end
      end

      describe 'failure' do
        it 'stores a datapoint' do
          subject.configuration.storage.should_receive(:update).with(event_configuration[:storage], 'event', :failure)
          subject.failure('event')
        end
      end
    end

    context 'when disabled' do
      before do
        subject.configuration.enabled = false
      end

      describe 'success' do
        it 'does nothing' do
          subject.configuration.storage.should_not_receive(:update)
          subject.success('event')
        end
      end

      describe 'failure' do
        it 'does nothing' do
          subject.configuration.storage.should_not_receive(:update)
          subject.failure('event')
        end
      end
    end
  end
end


spec/lib/imperfect/storage/cloudwatch_spec.rb
require 'spec_helper'

describe 'Imperfect::Storage::Cloudwatch' do
  subject do
    Imperfect::Storage::Cloudwatch.new(:secret_access_key => 'secret', :access_key_id => 'id')
  end

  describe '#update' do
    context 'with a success' do
      it 'submits the data to cloudwatch' do
        stub = stub_request(:post, 'https://monitoring.us-east-1.amazonaws.com/').
          with(:body => hash_including(
            'Namespace' => 'namespace',
            'MetricData.member.1.MetricName' => 'event',
            'MetricData.member.1.Value' => '0.0'
            )
          )

        subject.update({ :cloudwatch_namespace => 'namespace' }, 'event', :success)
        stub.should have_been_requested
      end
    end

    context 'with a failure' do
      it 'submits the data to cloudwatch' do
        stub = stub_request(:post, 'https://monitoring.us-east-1.amazonaws.com/').
          with(:body => hash_including(
            'Namespace' => 'namespace',
            'MetricData.member.1.MetricName' => 'event',
            'MetricData.member.1.Value' => '1.0'
            )
          )

        subject.update({ :cloudwatch_namespace => 'namespace' }, 'event', :failure)
        stub.should have_been_requested
      end
    end
  end

  describe 'error_rate' do
    it 'submits the data to cloudwatch' do
      success = """
      <GetMetricStatisticsResponse xmlns=\"http://monitoring.amazonaws.com/doc/2010-08-01/\">\n
        \ <GetMetricStatisticsResult>\n    <Datapoints>\n      <member>\n        <Timestamp>2012-12-11T19:00:00Z</Timestamp>\n
        \       <Unit>Count</Unit>\n        <Count>100.0</Count>\n      </member>\n    </Datapoints>\n
        \   <Label>event</Label>\n  </GetMetricStatisticsResult>\n
        \ <ResponseMetadata>\n    <RequestId>b94b3c91-43c5-11e2-be9b-799373c9f0f2</RequestId>\n
        \ </ResponseMetadata>\n</GetMetricStatisticsResponse>\n
      """
      failure = """
      <GetMetricStatisticsResponse xmlns=\"http://monitoring.amazonaws.com/doc/2010-08-01/\">\n
        \ <GetMetricStatisticsResult>\n    <Datapoints>\n      <member>\n        <Timestamp>2012-12-11T19:00:00Z</Timestamp>\n
        \       <Unit>Count</Unit>\n        <Sum>5.0</Sum>\n      </member>\n    </Datapoints>\n
        \   <Label>event</Label>\n  </GetMetricStatisticsResult>\n
        \ <ResponseMetadata>\n    <RequestId>b94b3c91-43c5-11e2-be9b-799373c9f0f2</RequestId>\n
        \ </ResponseMetadata>\n</GetMetricStatisticsResponse>\n
      """
      stub = stub_request(:post, 'https://monitoring.us-east-1.amazonaws.com/').to_return(:body => success).then.to_return(:body => failure)

      subject.error_rate({}, 'event', 300).should eq(0.05)
      stub.should have_been_requested.twice
    end
  end
end


lib/imperfect/storage/cloudwatch.rb
require 'aws-sdk'

module Imperfect::Storage
  class Cloudwatch
    def initialize(cloudwatch_configuration)
      @client = ::AWS::CloudWatch.new(cloudwatch_configuration)
    end

    # Public: Records the status of the given event.
    #
    # configuration - The Hash of configuration for this storage
    #                 method.
    # event_name    - The name of the event to be stored
    # status        - Either :success or :failure]
    #
    # Returns nothing.
    def update(configuration, event_name, status)
      metric_name = configuration[:cloudwatch_metric_name] || event_name
      namespace = configuration[:cloudwatch_namespace] || "Imperfect"

      value = status == :success ? 0 : 1

      @client.put_metric_data(
        :namespace => namespace,
        :metric_data => [{
          :metric_name => metric_name,
          :value => value
        }]
      )
    end

    # Protected: Returns the current error rate.
    #
    # configuration - The Hash of configuration for this storage
    #                 method.
    # event_name    - The name of the event to be stored
    # status        - Either :success or :failure
    # period        - The number of seconds to look back.
    def error_rate(configuration, event_name, period)
      total_count = count(configuration, event_name, :total, period)
      failure_count = count(configuration, event_name, :failure, period)

      failure_count/total_count.to_f
    end

  protected
    # Protected: Returns the number of instances of the event in the
    # given period of time.
    #
    # configuration - The Hash of configuration for this storage
    #                 method.
    # event_name    - The name of the event to be stored
    # type          - Either :total or :failure
    # period        - The number of seconds to look back.
    #
    # Returns the number of occurences of the specified event.
    def count(configuration, event_name, type, period)
      metric_name = configuration[:cloudwatch_metric_name] || event_name
      namespace = configuration[:cloudwatch_namespace] || "Imperfect"
      start_time = Time.now - period
      end_time = start_time + period

      metric = ::AWS::CloudWatch::Metric.new(namespace, metric_name, :config => @client.config)
      statistic = type == :total ? 'Count' : 'Sum'
      statistics = metric.statistics(
        :start_time => start_time,
        :end_time => end_time,
        :statistics => [statistic],
        :period => period
      )

      if statistics.first
        statistics.first[statistic.downcase.to_sym]
      else
        0
      end
    end
  end
end


lib/imperfect.rb
require 'imperfect/version'
require 'imperfect/storage'
require 'imperfect/alerts'

class Imperfect
  class << self
    attr_accessor :configuration
  end

  # Public: Allows the configuration of Imperfect using the given block.
  #
  # Example:
  #
  #  Imperfect.configure do |config|
  #    config.enabled = Rails.env.production?
  #
  #    config.enable_storage :cloudwatch, {
  #      :access_key_id => "your-aws-id",
  #      :secret_access_key => "your-aws-secret"
  #    }
  #    config.enable_alerts :pagerduty, {
  #      :api_key => "your-pagerduty-api-key"
  #    }
  #
  #    # Minimum time (in seconds) between re-triggering alerts
  #    config.minimum_alert_update_interval => 60
  #
  #    # All events that you would alert on must be specified here. Unconfigured
  #    # events will be silently ignored.
  #    config.events = {
  #      'event' => {
  #        # Alert if the failure rate exceeds this value.
  #        :acceptable_failure_rate => 0.10,
  #
  #        # The period of preceding time used to determine the current failure rate,
  #        # defaults to 300s.
  #        :lookback_period => 300,
  #
  #        :alerts => {
  #          :pagerduty => {
  #            # The name of the pagerduty service to mark as failed.
  #            :service_name => 'service-event',
  #          }
  #        },
  #
  #        :storage => {
  #          # The group which this metric should belong to on cloudwatch.
  #          :cloudwatch_namespace => 'Imperfect',
  #          # The name of the cloudwatch metric to use.
  #          :cloudwatch_metric_name => 'event-success',
  #        }
  #      }
  #    }
  #  end
  #
  # Returns nothing.
  def self.configure
    self.configuration ||= Configuration.new
    yield(configuration)
  end

  # Public: Track a successful completion of the given event.
  #
  # event - The name of the successful event.
  #
  # Returns nothing.
  def self.success(event)
    return unless configuration.enabled

    event_configuration = configuration.events[event]
    return unless event_configuration

    configuration.storage.update(event_configuration[:storage], event, :success)
  end

  # Public: Track a unsuccessful completion of the given event.
  #
  # event - The name of the successful event.
  #
  # Returns nothing.
  def self.failure(event)
    return unless configuration.enabled

    event_configuration = configuration.events[event]
    return unless event_configuration

    configuration.storage.update(event_configuration[:storage], event, :failure)
  end

  # Test-only: This function is used to test whether an alerting
  # configuration works as expected.
  #
  # Returns nothing.
  def self.trigger_alert!(event)
    return unless configuration.enabled
  end

  class Configuration
    attr_accessor :minimum_alert_update_interval, :events, :enabled
    attr_reader :storage, :alerts

    def initialize
      @minimum_alert_update_interval = 60
      @enabled = true
      @events = {}
    end

    def enable_storage(type, configuration)
      @storage = case type
      when :cloudwatch
        Imperfect::Storage::Cloudwatch.new(configuration)
      else
        raise "Invalid storage type: #{type}"
      end
    end

    def enable_alerts(type, configuration)
      @alerts = case type
      when :pagerduty
        Imperfect::Alerts::Pagerduty.new(configuration)
      else
        raise "Invalid alert type: #{type}"
      end
    end
  end
end
